library/core/src/num/mod.rs

```diff
@@ -846,37 +870,66 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+    if radix <= 16 && src.len() <= mem::size_of::<T>() * 2 - if is_signed_ty { 1 } else { 0 } {
```

Comment:
Why does this multiply by two?

Reply:
I believe that is correct but he didn't write it down. `u8` max out by `ff` which is 2 characters, `u8` max out by `ffff` which is 4 characters, maybe it would be better to leave a comment here instead of confusing the reader.

I believe this only applies to radix of 16, for instance radix of 8 should have `len() == size_of::<T>()` so not sure if the compiler can optimized that out.

Reply:
Yeah comment sorely needed here - sorry!
We could have a look up table T,radix=>safe_length - that would make this optimal. At the moment this rule of * 2 is sufficient but way too conservative for lower radix, it's ok for radix 10 but could be better. I was assuming 99% of the time radix = 10 or 16.

I guess we could add an array of 36 to each type for safe_len as part of the macro. I presume those arrays would only be referenced at compile time and would get thrown away rather than add to the final binary size?

Reply:
@pickfire actually it's the other way around: for radix of 8 should have len() <= size_of::<T>() * 4 because you can fit twice as many radix 8 chars in the type as radix 16 chars. (`8888` would be equiv to `ff`)

Reply:
@gilescope That's not accurate. One radix 8 character represents 3 bits, while one radix 16 character represents 4 bits; that's not a factor of 2 difference. I also think this optimization doesn't need to be incredibly precise; it's just intended to cover the common case. Having just the simple case of multiplying by 2 seems reasonable, and larger decimal numbers just won't get the optimization.

Reply:
`8888` should have been `7777` for radix 8 (my bad).
`7777` is 4095 which is bigger than `FF`'s 255. Agreed that doesn't hold at all, will stick at * 2.

Reply:
I'm going to resolve this one - let me know if I can make the comment any clearer.